### DISCUSSION TITLE
0.9.x version

### DISCUSSION THREAD

**Discussion by brichet** [jupyter/nbgrader#1749] [General]
There are currently 2 maintained versions of `nbgrader`
- **0.7.x** with support for `notebook<7` extensions on branch [0.7.x](https://github.com/jupyter/nbgrader/tree/0.7.x)
- **0.8.x** with support for `notebook<7` and `jupyterlab<4` extensions on branch [main](https://github.com/jupyter/nbgrader/tree/main)

There is also this [draft PR](https://github.com/jupyter/nbgrader/pull/1684) to extend support to `jupyterlab==4` and `notebook==7`. This PR grows until Jupyterlab 4 and Notebook 7 are released.

I propose here to stop development on **0.7.x** version, since its features should be covered by **0.8.x** version, and to start a new **0.9.x** version to support the next versions of Jupyterlab and Notebook.

**NOTE**
It might be interesting to investigate a way to independently install the JupyterLab or Notebook extensions (and their dependencies) in version **0.8.x**. Some people may still want to avoid the `nodejs` installation for the classic Notebook.

cc. @jhamrick @SylvainCorlay @jtpio

**Comment by jtpio** [comment]
Thanks @brichet for starting this discussion :+1: 

> There is also this [draft PR](https://github.com/jupyter/nbgrader/pull/168) to extend support to `jupyterlab==4` and `notebook==7`. This PR grows until Jupyterlab 4 and Notebook 7 are released.

Right, it would make sense to create a `0.8.x` branch, and then target the next `0.9.x` version on `main` directly. This would allow for merging this PR and be able to iterate on it more easily.

**Comment by brichet** [comment]
<h3>0.8.x branch is created.</h3>

The **0.8.x** branch will keep compatibility with *Jupyterlab 3.x* and *Notebook 6.x*.

The **main** will be used for compatibility with *Jupyterlab 4* and *Notebook 7*, and should release versions `0.9.x`.

**Comment by Mad31** [comment]
Hi ?
a date for the release 9.X to have à jupyterlab 4 compatibility ?

**Reply by brichet** [reply]
A pre-release is out: https://github.com/jupyter/nbgrader/releases/tag/v0.9.0a0
It can be installed with `pip install --pre nbgrader`

**Comment by brichet** [comment]
No date for now, we have to wait for:
- Notebook 7 to be released, before the end of the month I believe.
- this [PR](https://github.com/jupyter/nbgrader/pull/1684) to be reviewed and merged. Feel free to test it and comment it.

**Comment by jtpio** [comment]
Now that https://github.com/jupyter/nbgrader/pull/1684 has been merged into `main`, we should be able to make the first `0.9.0a0` pre-release.

Ad it looks like the `0.8.x` branch already exists :+1: https://github.com/jupyter/nbgrader/tree/0.8.x

**Reply by Mad31** [reply]
I've tried v0.9.0a0  👍 
```
JupyterLab v4.0.3
/usr/local/share/jupyter/labextensions
        jupyterlab_pygments v0.2.2 enabled  X (python, jupyterlab_pygments)
        jupyterlab-myst v2.0.1 enabled OK (python, jupyterlab_myst)
        ipycanvas v0.13.1 enabled OK
        nbgrader v0.9.0-alpha.0 enabled  X (python, nbgrader)
        @jupyter-widgets/jupyterlab-manager v5.0.8 enabled OK (python, jupyterlab_widgets)
        @jupyterlab/latex v3.1.0 enabled  X (python, jupyterlab-latex)

```
but 

```
The following extensions are outdated:
        jupyterlab_pygments
        nbgrader
        @jupyterlab/latex
```

and i can't disable extension 'formgrader' for example for student

**Reply by brichet** [reply]
Thanks for the feedback.
Can you open an issue about that ?

**Comment by trevorcampbell** [comment]
@brichet @jtpio any chance we can get a few of the outstanding bugfix PRs included in 0.9.0? These would be a big improvement in using `nbgrader` in larger classes... I'm specifically thinking of #1770, but @tuncbkose and some others have a few open PRs that seem like they'd be helpful as well.

I would be happy to contribute any testing effort needed, with a bit of guidance. 

( I also have the two autotest PRs out #1650 and #1656, and we've been using that code very successfully in our class for a few semesters now. But I imagine those will involve a biiiit more work to merge :-) )

**Reply by trevorcampbell** [reply]
(hopefully this is the right place to discuss that -- I couldn't really tell if the 0.9.x thread was for discussing a new 0.9.0 release or something else -- apologies if not!)

**Reply by brichet** [reply]
Hey @trevorcampbell.
Version 0.9.0 focuses mainly on compatibility with Jupyterlab 4 and Notebook 7.
Nevertheless, I agree that some enhancement and bugfix PRs should be reviewed and merged. It seems that few maintainers are able to spend time on `nbgrader` at the moment.

I hope to be able to take a look at these PRs in a few weeks (I won't be available before then).

**Comment by SylvainCorlay** [comment]
Following today's discussion, we should probably make another pre-release of nbgrader 0.9 that uses Notebook 7 final instead of the release candidate.

**Reply by brichet** [reply]
I opened https://github.com/jupyter/nbgrader/pull/1805

**Reply by brichet** [reply]
It's out: https://github.com/jupyter/nbgrader/releases/tag/v0.9.0a1

**Comment by lahwaacz** [comment]
I have [selectively disabled](https://nbgrader.readthedocs.io/en/stable/user_guide/installation.html) all nbgrader lab extensions and server extensions, but the Nbgrader context menu in JupyterLab still has active buttons:

![screenshot-2023-08-04@22:49:28](https://github.com/jupyter/nbgrader/assets/1289205/c30e94a8-f66f-4002-ae2f-2619463a9921)

Notice that only "Create assignment" is grayed out. Shouldn't the other be the same?

My `~/.jupyter/labconfig/page_config.json` contains:
```json
{
  "disabledExtensions": {
    "nbgrader/formgrader": true,
    "nbgrader/assignment-list": true,
    "nbgrader/course-list": true,
    "nbgrader/create-assignment": true,
    "nbgrader/validate-assignment": true
  }
}
```

Then `~/.jupyter/jupyter_server_config.d/` contains four `.json` files that disable the corresponding server extensions, but this is not relevant since this is the JupyterLab interface.

**Reply by lahwaacz** [reply]
I managed to figure out this issue. The problem is that the [documentation](https://nbgrader.readthedocs.io/en/stable/user_guide/installation.html) is wrong about the labextension plugin IDs in many places: it's not `nbgrader/assignment-list` or even `nbgrader/assignment_list`, but `nbgrader:assignment-list`, and similar for the others. This is pretty insidious since running `jupyter labextension disable` with the wrong plugin ID does not result in error and `jupyter labextension list` also does not help (it lists only `nbgrader` unless some of its labextensions is already disabled). I submitted https://github.com/jupyter/nbgrader/pull/1811 with a fix.

**Comment by trevorcampbell** [comment]
@brichet I have been working on rebasing our AutoTest PR on 0.9.x, and I ran into this (minor) issue #1814 with the current `main` branch that was introduced recently. I've made a PR to fix it #1815 . Very minor change. Also, perhaps you want to `del` the unused `args` as well.

**Comment by trevorcampbell** [comment]
@brichet another possibly relevant issue: #1816 . The CI on `ui-tests` looks like it's passing, but the logs contain many uncaught errors (mostly 404 / file not found, but a few "unwriteable directory" etc as well).

**Reply by trevorcampbell** [reply]
After digging into it a little bit more, these errors have been appearing in the `ui-tests` for a very long time (at least as far back as GitHub maintains action logs). 

It looks related to the setup of the test directory / when the nbgrader config file is loaded... things aren't in the right place, and nbgrader isn't respecting the `nb_exchange_test` directory that gets created in `beforeEach` [here](https://github.com/jupyter/nbgrader/blob/af47e5eaf880d456f79030e68c0feb0de1fb6313/nbgrader/tests/ui-tests/assignment_list.spec.ts#L71C8-L71C8). It's trying to use the exchange at `/usr/local/share/nbgrader/exchange`.

Unfortunately I don't know enough about javascript / galata to help fix the issue...

**Reply by trevorcampbell** [reply]
The new AutoTest code is ready and compatible with 0.9.x (available here https://github.com/ubc-dsci/nbgrader/tree/autotest-0.9.x ). This branch supercedes the current open PR at #1650 , which was too out of date to rebase without a lot of headache.

It passes all tests except `ui-tests`. I don't make any modifications to the UI, but `ui-tests` runs `generate_assignment` which triggers AutoTest. I don't think I'll be able to figure the rest out until the above issues in `ui-tests` are handled.

**Reply by trevorcampbell** [reply]
I decided just to remove the new notebooks from the UI tests to avoid triggering AutoTest there. AutoTest doesn't have anything to do with the UI, so I don't think this needs to be included in those tests anyway.

PR opened at #1817
